lib/membrane_rtmp_plugin/rtmp/source/source.ex

```diff
@@ -1,148 +1,155 @@
+
+    send(self(), :start_receiving)
+
+    actions = [
+      caps: {:output, %Membrane.RemoteStream{content_format: Membrane.FLV, type: :bytestream}}
+    ]
+
+    {{:ok, actions}, %{state | receiver_pid: receiver_process}}
   end
 
-  @impl true
-  def handle_demand(type, _size, _unit, _ctx, %{stale_frame: {type, buffer}} = state) do
-    # There is stale frame, which indicates that that the source was blocked waiting for demand from one of the outputs
-    # It now arrived, so we request next frame and output the one that blocked us
-    send(state.provider, :get_frame)
-    {{:ok, buffer: {type, buffer}}, %{state | stale_frame: nil}}
+  defp receive_loop(socket, target) do
+    receive do
+      :need_more_data ->
+        :ok = :inet.setopts(socket, active: :once)
+
+      {:tcp, _port, packet} ->
+        send(target, {:tcp, socket, packet})
+
+      {:tcp_closed, _port} ->
+        send(target, {:tcp_closed, socket})
+
+      :terminate ->
+        exit(:normal)
+
+      _message ->
+        :noop
+    end
+
+    receive_loop(socket, target)
   end
 
   @impl true
-  def handle_demand(_type, _size, _unit, _ctx, state) do
+  def handle_demand(_pad, _size, _unit, _ctx, state) when state.socket_ready? do
+    send(state.receiver_pid, :need_more_data)
     {:ok, state}
   end
 
   @impl true
-  def handle_other({Native, :format_info_ready, native_ref}, _ctx, state) do
-    actions = get_format_info_actions(native_ref)
-    {{:ok, actions}, state}
+  def handle_demand(_pad, _size, _unit, _ctx, state) do
+    {:ok, state}
   end
 
   @impl true
-  def handle_other({Native, :read_frame, {:ok, type, pts, dts, frame}}, ctx, state)
-      when ctx.playback_state == :playing do
-    pts = Time.milliseconds(pts)
-    dts = Time.milliseconds(dts)
-
-    buffer = %Buffer{
-      pts: pts,
-      dts: dts,
-      payload: prepare_payload(type, frame)
-    }
-
-    if get_in(ctx.pads, [type, :demand]) > 0 do
-      send(state.provider, :get_frame)
-      {{:ok, buffer: {type, buffer}}, state}
-    else
-      # if there is no demand for element of this type so we wait until it appears
-      # effectively, it results in source adapting to the slower of the two outputs
-      {:ok, %{state | stale_frame: {type, buffer}}}
-    end
+  def handle_playing_to_prepared(_ctx, state) do
+    send(state.receiver_pid, :terminate)
+    Process.unlink(state.receiver_pid)
+    {:ok, %{state | receiver_pid: nil}}
   end
 
   @impl true
-  def handle_other({Native, :read_frame, :end_of_stream}, _ctx, state) do
-    Membrane.Logger.debug("Received end of stream")
-    {{:ok, end_of_stream: :audio, end_of_stream: :video}, state}
+  def handle_other(:start_receiving, _ctx, state) do
+    case :gen_tcp.controlling_process(state.socket, state.receiver_pid) do
+      :ok ->
+        :ok = :inet.setopts(state.socket, active: :once)
+        {:ok, %{state | socket_ready?: true}}
+
+      {:error, :not_owner} ->
+        Process.send_after(self(), :start_receiving, 200)
```

Comment:
Is it possible that we will never become the controlling process therefore we will be stuck in this function? We should probably give up trying after the first retry.

Reply:
Ok, with retrying after 200ms a single retry wasn't enough to always succeed - I decided for 3 retries.